test/Ocean/shallow_water/GyreDriver.jl

```diff
@@ -0,0 +1,220 @@
+
+  return nothing
+end
+
+#########################
+# Timestepping function #
+#########################
+
+function run(mpicomm, topl, ArrayType, N, dt, DFloat, model, test)
+  grid = DiscontinuousSpectralElementGrid(topl,
+                                          FloatType = DFloat,
+                                          DeviceArray = ArrayType,
+                                          polynomialorder = N,
+                                         )
+
+  dg = DGModel(model,
+               grid,
+               Rusanov(),
+               CentralNumericalFluxDiffusive(),
+               CentralGradPenalty())
+
+  param = init_ode_param(dg)
+  Q  = init_ode_state(dg, param, DFloat(0))
+  Qe = init_ode_state(dg, param, DFloat(timeend))
+
+  lsrk = LSRK144NiegemannDiehlBusch(dg, Q; dt = dt, t0 = 0)
+
+  cb = ()
+
+  if test > 2
+    outprefix = @sprintf("ic_mpirank%04d_ic", MPI.Comm_rank(mpicomm))
+    statenames = flattenednames(vars_state(model, eltype(Q)))
+    auxnames = flattenednames(vars_aux(model, eltype(Q)))
+    writevtk(outprefix, Q, dg, statenames, param.aux, auxnames)
+
+    outprefix = @sprintf("exact_mpirank%04d", MPI.Comm_rank(mpicomm))
+    statenames = flattenednames(vars_state(model, eltype(Qe)))
+    auxnames = flattenednames(vars_aux(model, eltype(Qe)))
+    writevtk(outprefix, Qe, dg, statenames, param.aux, auxnames)
+
+    step = [0]
+    vtkpath = outname
+    mkpath(vtkpath)
+    cbvtk = GenericCallbacks.EveryXSimulationSteps(1000)  do (init=false)
+      outprefix = @sprintf("%s/mpirank%04d_step%04d",vtkpath,
+                           MPI.Comm_rank(mpicomm), step[1])
+      @debug "doing VTK output" outprefix
+      statenames = flattenednames(vars_state(model, eltype(Q)))
+      auxnames = flattenednames(vars_aux(model, eltype(Q)))
+      writevtk(outprefix, Q, dg, statenames, param.aux, auxnames)
+      step[1] += 1
+      nothing
+    end
+    cb = (cb..., cbvtk)
+  end
+
+  solve!(Q, lsrk, param; timeend=timeend, callbacks=cb)
+
+  error = euclidean_distance(Q, Qe) / norm(Qe)
+  @info @sprintf("""Finished
+                 error = %.16e
+                 """, error)
+
+  return error
+end
+
+
+################
+# Start Driver #
+################
+
+let
+  MPI.Initialized() || MPI.Init()
+  mpicomm = MPI.COMM_WORLD
+
+  ll = uppercase(get(ENV, "JULIA_LOG_LEVEL", "INFO"))
+  loglevel = ll == "DEBUG" ? Logging.Debug :
+  ll == "WARN"  ? Logging.Warn  :
+  ll == "ERROR" ? Logging.Error : Logging.Info
+  logger_stream = MPI.Comm_rank(mpicomm) == 0 ? stderr : devnull
+  global_logger(ConsoleLogger(logger_stream, loglevel))
+  @static if haspkg("CUDAnative")
+    device!(MPI.Comm_rank(mpicomm) % length(devices()))
+  end
+
+  @static if haspkg("CuArrays")
+    ArrayType = CuArray
+  else
+    ArrayType = Array
+  end
+
+  model = setup_model(DFloat, stommel, linear, τₒ, fₒ, β, λ, ν, Lˣ, Lʸ, H)
+
+  if test == 1
+    cellsrange = 10:10
+    orderrange = 4:4
+    testval = 1.6068814534535144e-03
```

Comment:
Have you tested it with more than one level (i.e., this number is actually an error not an not just a random number)?

Reply:
We ran with 4 levels, 2x2 in nelems and polyorder, but the code for doing the comparisons between levels failed, so the current test is jut checking the error for nelems = 10 and polyorder = 4

Reply:
Great. Just wanted to check.